src/nrnoc/multicore.cpp

```diff
@@ -1313,54 +1266,11 @@ int nrn_allow_busywait(int b) {
     return old;
 }
 
-#if USE_PTHREAD
-static long waste_;
-static void* waste(void* v) {
-    size_t i, j, n;
-    n = (size_t) v;
-    j = 0;
-    for (i = 0; i < n; ++i) {
-        j += i;
-    }
-    /* hoping it is  not optimized away */
-    waste_ = j;
-    return nullptr;
-}
-
-#define _nt_ 32
-static double trial(int ip) {
-    int i;
-    double t;
-    pthread_t* th;
-    th = (pthread_t*) ecalloc(ip, sizeof(pthread_t));
-    t = nrn_timeus();
-    for (i = 0; i < ip; ++i) {
-        pthread_create(th + i, nullptr, waste, (void*) 100000000);
-    }
-    for (i = 0; i < ip; ++i) {
-        pthread_join(th[i], nullptr);
-    }
-    t = nrn_timeus() - t;
-    free((char*) th);
-    return t;
-}
-#endif
-
 int nrn_how_many_processors() {
-#if USE_PTHREAD
-    int i, ip;
-    double t1, t2;
-    printf("nthread walltime (count to 1e8 on each thread)\n");
-    t1 = trial(1);
-    printf("%4d\t %g\n", 1, t1);
-    for (ip = 2; ip <= _nt_; ip *= 2) {
-        t2 = trial(ip);
-        printf("%4d\t %g\n", ip, t2);
-        if (t2 > 1.3 * t1) {
-            return ip / 2;
-        }
-    }
-    return _nt_;
+#if NRN_ENABLE_THREADS
+    // For machines with hyperthreading this probably returns the number of
+    // logical, not physical, cores.
+    return std::thread::hardware_concurrency();
```

Comment:
https://nrn.readthedocs.io/en/latest/hoc/modelspec/programmatic/network/parcon.html?highlight=thread#ParallelContext.thread_how_many_proc
needs updating.

Reply:
OK, I'll do that. Note that the old code was returning a wrong, small number (typically 1 or 2, I think) on the (36-core) test machine I was using.

Reply:
It seems a bit unfortunate that the documentation is not shared between HOC and Python: https://github.com/neuronsimulator/nrn/blob/e36c04916f3746a1821d79d2d51d5e677c7e39ca/docs/hoc/modelspec/programmatic/network/parcon.rst and https://github.com/neuronsimulator/nrn/blob/e36c04916f3746a1821d79d2d51d5e677c7e39ca/docs/python/modelspec/programmatic/network/parcon.rst.

Reply:
That's interesting. When there is no memory being used, I would have expected 36 cores would give a 36 fold speedup.
But I guess that assumes nothing else is contending for cycles. Am I correct in suspecting that one (a simulation) will never do better than the old code, and likely worse as the memory latency/bandwidth will never be able to supply the cores with something to do?

Reply:
I didn't investigate at the time as I assumed the optimiser was defeating the old code, which seems to be correct:
```console
$ clang++ -o foo foo.cpp -pthread -O0
$ ./foo
nthread walltime (count to 1e8 on each thread)
   1     0.297978
   2     0.298184
   4     0.298539
   8     0.298533
  16     0.298732
  32     0.344102
  64     0.420679
32
$ clang++ -o foo foo.cpp -pthread -O1
$ ./foo
nthread walltime (count to 1e8 on each thread)
   1     0.000100851
   2     0.000100851
   4     0.000224829
2
```
sometimes the optimised version returns 1.

At least on the test system, the `std::thread::hardware_concurrency()` version seems to be unaware of any pinning to a reduced number of cores:
```console
$ clang++ -o foo foo.cpp -pthread -O0
$ srun -n 1 -c 8 --exact ./foo
nthread walltime (count to 1e8 on each thread)
   1     0.298106
   2     0.29824
   4     0.344123
   8     0.358102
  16     0.763359
8
std::thread::hardware_concurrency(): 72
```
where the old code gives a "better" answer...but I don't think this is really relevant given that you have to completely disable optimisation to get that "better" answer.

Reply:
At any rate. Let's not let this affect the merge of this PR. The only reason the old existed was to avoid the uselessness of counting hyperthreads.